File: recipes/subms-lsm-tree/java/src/main/java/com/submillisecond/recipes/lsm/SSTable.java
package com.submillisecond.recipes.lsm;

import java.io.BufferedOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import com.submillisecond.recipes.bloom.BloomFilter;

/**
 * Immutable on-disk sorted run, slurped into memory on {@link #open}.
 *
 * <p>File layout (big-endian):
 * <pre>
 *   records: (key_len:u32 key:utf-8 flag:u8 value_len:u32 value:bytes)*
 *   bloom:   bit_count:u32 k:u32 word_count:u32 (word:u64)*
 *   footer:  records_end_offset:u64 magic:u32
 *   flag := 0x00 (present) | 0x01 (tombstone, value_len == 0)
 *   magic := 0x4C534D54 ("LSMT")
 * </pre>
 *
 * On open the whole file is read into a byte buffer, the bloom filter is parsed
 * out of the trailer, and reads operate entirely against memory. A get short-circuits
 * on a bloom miss.
 */
final class SSTable {

    private static final int  MAGIC          = 0x4C534D54;
    private static final byte FLAG_VALUE     = 0x00;
    private static final byte FLAG_TOMBSTONE = 0x01;
    private static final int  FOOTER_BYTES   = 8 + 4;

    private final Path path;
    private final byte[] buf;
    private final int recordsEnd;
    private final BloomFilter bloom;
    /**
     * Start byte offset of each record, in key order (records are stored
     * sorted). Built once on {@code open} from the slurped buffer; lets
     * {@code get} and {@code range} binary-search to a key instead of scanning
     * from the file start. Costs one int per record - the on-disk format is
     * unchanged.
     */
    private final int[] offsets;

    private SSTable(Path path, byte[] buf, int recordsEnd, BloomFilter bloom) {
        this.path = path;
        this.buf = buf;
        this.recordsEnd = recordsEnd;
        this.bloom = bloom;
        this.offsets = indexOffsets(buf, recordsEnd);
    }

    static SSTable open(Path path) throws IOException {
        byte[] buf = Files.readAllBytes(path);
        if (buf.length < FOOTER_BYTES) throw new IOException("SSTable too small: " + path);

        int magicOff = buf.length - 4;
        int magic = readIntBE(buf, magicOff);
        if (magic != MAGIC) throw new IOException("bad SSTable magic in " + path);

        long recordsEndLong = readLongBE(buf, buf.length - FOOTER_BYTES);
        if (recordsEndLong < 0 || recordsEndLong > buf.length - FOOTER_BYTES) {
            throw new IOException("bad records_end offset in " + path);
        }
        int recordsEnd = (int) recordsEndLong;
        int bloomLen = buf.length - FOOTER_BYTES - recordsEnd;
        BloomFilter bloom = BloomFilter.parse(buf, recordsEnd, bloomLen);
        return new SSTable(path, buf, recordsEnd, bloom);
    }

    /** One pass over the sorted records collecting each record's start offset. */
    private static int[] indexOffsets(byte[] buf, int recordsEnd) {
        int count = 0;
        int p = 0;
        while (p < recordsEnd) {
            count++;
            int keyLen = readIntBE(buf, p);
            p += 4 + keyLen + 1;
            int valueLen = readIntBE(buf, p);
            p += 4 + valueLen;
        }
        int[] offs = new int[count];
        p = 0;
        for (int i = 0; i < count; i++) {
            offs[i] = p;
            int keyLen = readIntBE(buf, p);
            p += 4 + keyLen + 1;
            int valueLen = readIntBE(buf, p);
            p += 4 + valueLen;
        }
        return offs;
    }

    /** Compare the key of the record at {@code off} against {@code keyBytes}. */
    private int compareKeyAt(int off, byte[] keyBytes) {
        int keyLen = readIntBE(buf, off);
        return compareKey(buf, off + 4, keyLen, keyBytes);
    }

    static SSTable write(
        Path path,
        int expectedEntries,
        Iterable<Map.Entry<String, byte[]>> sortedEntries
    ) throws IOException {
        BloomFilter bloom = new BloomFilter(expectedEntries);
        long recordsEnd = 0;
        try (DataOutputStream out = new DataOutputStream(new BufferedOutputStream(Files.newOutputStream(path)))) {
            for (Map.Entry<String, byte[]> e : sortedEntries) {
                bloom.add(e.getKey());
                byte[] keyBytes = e.getKey().getBytes(StandardCharsets.UTF_8);
                byte[] value = e.getValue();
                out.writeInt(keyBytes.length);
                out.write(keyBytes);
                if (value == null) {
                    out.writeByte(FLAG_TOMBSTONE);
                    out.writeInt(0);
                    recordsEnd += 4L + keyBytes.length + 1 + 4;
                } else {
                    out.writeByte(FLAG_VALUE);
                    out.writeInt(value.length);
                    out.write(value);
                    recordsEnd += 4L + keyBytes.length + 1 + 4 + value.length;
                }
            }
            bloom.writeTo(out);
            out.writeLong(recordsEnd);
            out.writeInt(MAGIC);
        }
        return open(path);
    }

    /**
     * @return {@code Optional.empty()} if the key is not in this run;
     *         a hit carrying {@code null} value if it is tombstoned;
     *         a hit carrying the value otherwise.
     *
     *         {@code checkBloom = false} skips the bloom probe and goes
     *         straight to the scan - used by {@link BloomMode#OFF} to
     *         measure the optimisation's value.
     */
    Optional<Hit> get(String key, boolean checkBloom) {
        if (checkBloom && !bloom.mightContain(key)) return Optional.empty();

        byte[] keyBytes = key.getBytes(StandardCharsets.UTF_8);
        // Records are sorted, so binary-search the offset index (O(log n)) rather
        // than scan from the file start.
        int lo = 0, hi = offsets.length - 1;
        while (lo <= hi) {
            int mid = (lo + hi) >>> 1;
            int off = offsets[mid];
            int cmp = compareKeyAt(off, keyBytes);
            if (cmp == 0) {
                int p = off + 4 + readIntBE(buf, off);
                byte flag = buf[p]; p += 1;
                int valueLen = readIntBE(buf, p); p += 4;
                if (flag == FLAG_TOMBSTONE) return Optional.of(new Hit(null));
                byte[] value = new byte[valueLen];
                System.arraycopy(buf, p, value, 0, valueLen);
                return Optional.of(new Hit(value));
            }
            if (cmp < 0) lo = mid + 1; else hi = mid - 1;
        }
        return Optional.empty();
    }

    /**
     * Records whose key is in {@code [lo, hi)} (a {@code null} bound is
     * unbounded), in key order, as {@code (key, value)} entries - a tombstone
     * surfaces as an entry with a {@code null} value. Binary-searches the offset
     * index to seek to {@code lo}, then scans forward only across the window.
     */
    List<Map.Entry<String, byte[]>> range(String lo, String hi) {
        byte[] loBytes = lo == null ? null : lo.getBytes(StandardCharsets.UTF_8);
        byte[] hiBytes = hi == null ? null : hi.getBytes(StandardCharsets.UTF_8);
        // Lower bound: first record whose key is >= lo.
        int start = 0;
        if (loBytes != null) {
            int a = 0, b = offsets.length;
            while (a < b) {
                int mid = (a + b) >>> 1;
                if (compareKeyAt(offsets[mid], loBytes) < 0) a = mid + 1;
                else b = mid;
            }
            start = a;
        }
        List<Map.Entry<String, byte[]>> out = new ArrayList<>();
        for (int i = start; i < offsets.length; i++) {
            int off = offsets[i];
            int keyLen = readIntBE(buf, off);
            int keyOff = off + 4;
            if (hiBytes != null && compareKey(buf, keyOff, keyLen, hiBytes) >= 0) break;
            int p = keyOff + keyLen;
            byte flag = buf[p]; p += 1;
            int valueLen = readIntBE(buf, p); p += 4;
            String key = new String(buf, keyOff, keyLen, StandardCharsets.UTF_8);
            byte[] value = null;
            if (flag != FLAG_TOMBSTONE) {
                value = new byte[valueLen];
                System.arraycopy(buf, p, value, 0, valueLen);
            }
            out.add(new AbstractMap.SimpleImmutableEntry<>(key, value));
        }
        return out;
    }

    Path path() {
        return path;
    }

    private static int compareKey(byte[] buf, int off, int len, byte[] keyBytes) {
        int n = Math.min(len, keyBytes.length);
        for (int i = 0; i < n; i++) {
            int a = buf[off + i] & 0xff;
            int b = keyBytes[i] & 0xff;
            if (a != b) return a - b;
        }
        return Integer.compare(len, keyBytes.length);
    }

    private static int readIntBE(byte[] b, int o) {
        return ((b[o] & 0xff) << 24)
             | ((b[o + 1] & 0xff) << 16)
             | ((b[o + 2] & 0xff) << 8)
             |  (b[o + 3] & 0xff);
    }

    private static long readLongBE(byte[] b, int o) {
        long hi = (long) readIntBE(b, o) & 0xffffffffL;
        long lo = (long) readIntBE(b, o + 4) & 0xffffffffL;
        return (hi << 32) | lo;
    }

    record Hit(byte[] value) {
        boolean isTombstone() {
            return value == null;
        }
    }
}

File: recipes/subms-lsm-tree/rust/src/sstable.rs
//! Immutable on-disk sorted run, slurped into memory on [`SsTable::open`].
//!
//! File layout (big-endian):
//! ```text
//! records: (key_len:u32 key:utf-8 flag:u8 value_len:u32 value:bytes)*
//! bloom:   bit_count:u32 k:u32 word_count:u32 (word:u64)*
//! footer:  records_end_offset:u64 magic:u32
//! flag := 0x00 (present) | 0x01 (tombstone, value_len == 0)
//! magic := 0x4C534D54 ("LSMT")
//! ```
//!
//! On open the whole file is read into a `Vec<u8>`, the bloom filter is
//! parsed out of the trailer, and reads operate entirely against memory.
//! A get short-circuits on a bloom miss.

use std::fs;
use std::io::{self, BufWriter, Write};
use std::path::{Path, PathBuf};

use subms_bloom_filter::BloomFilter;

const MAGIC: u32 = 0x4C534D54;
const FLAG_VALUE: u8 = 0x00;
const FLAG_TOMBSTONE: u8 = 0x01;
const FOOTER_BYTES: usize = 8 + 4;

pub(crate) struct SsTable {
    #[allow(dead_code)]
    path: PathBuf,
    buf: Vec<u8>,
    records_end: usize,
    bloom: BloomFilter,
    /// Start byte offset of each record, in key order (records are stored
    /// sorted). Built once on `open` from the slurped buffer; lets `range`
    /// binary-search to `lo` instead of scanning from the file start. Costs one
    /// `usize` per record - the on-disk format is unchanged.
    offsets: Vec<usize>,
}

impl SsTable {
    pub(crate) fn open(path: impl Into<PathBuf>) -> io::Result<Self> {
        let path = path.into();
        let buf = fs::read(&path)?;
        if buf.len() < FOOTER_BYTES {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "sstable too small",
            ));
        }
        let magic_off = buf.len() - 4;
        let magic = u32::from_be_bytes(buf[magic_off..magic_off + 4].try_into().unwrap());
        if magic != MAGIC {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "bad sstable magic",
            ));
        }
        let footer_off = buf.len() - FOOTER_BYTES;
        let records_end =
            u64::from_be_bytes(buf[footer_off..footer_off + 8].try_into().unwrap()) as usize;
        if records_end > footer_off {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "bad records_end offset",
            ));
        }
        let bloom = BloomFilter::parse(&buf[records_end..footer_off])?;
        let offsets = Self::index_offsets(&buf, records_end);
        Ok(Self {
            path,
            buf,
            records_end,
            bloom,
            offsets,
        })
    }

    /// One pass over the sorted records collecting each record's start offset.
    fn index_offsets(buf: &[u8], records_end: usize) -> Vec<usize> {
        let mut offsets = Vec::new();
        let mut p = 0usize;
        while p < records_end {
            offsets.push(p);
            let key_len = u32::from_be_bytes(buf[p..p + 4].try_into().unwrap()) as usize;
            p += 4 + key_len + 1;
            let value_len = u32::from_be_bytes(buf[p..p + 4].try_into().unwrap()) as usize;
            p += 4 + value_len;
        }
        offsets
    }

    /// Key bytes of the record starting at `off`.
    fn key_at(&self, off: usize) -> &[u8] {
        let key_len = u32::from_be_bytes(self.buf[off..off + 4].try_into().unwrap()) as usize;
        &self.buf[off + 4..off + 4 + key_len]
    }

    pub(crate) fn write<'a, I>(
        path: impl Into<PathBuf>,
        expected_entries: usize,
        sorted_entries: I,
    ) -> io::Result<Self>
    where
        I: IntoIterator<Item = (&'a str, Option<&'a [u8]>)>,
    {
        let path = path.into();
        let mut bloom = BloomFilter::new(expected_entries);
        let mut records_end: u64 = 0;
        {
            let mut out = BufWriter::new(fs::File::create(&path)?);
            for (key, value) in sorted_entries {
                bloom.add(key);
                let kb = key.as_bytes();
                out.write_all(&(kb.len() as u32).to_be_bytes())?;
                out.write_all(kb)?;
                match value {
                    Some(v) => {
                        out.write_all(&[FLAG_VALUE])?;
                        out.write_all(&(v.len() as u32).to_be_bytes())?;
                        out.write_all(v)?;
                        records_end += (4 + kb.len() + 1 + 4 + v.len()) as u64;
                    }
                    None => {
                        out.write_all(&[FLAG_TOMBSTONE])?;
                        out.write_all(&0u32.to_be_bytes())?;
                        records_end += (4 + kb.len() + 1 + 4) as u64;
                    }
                }
            }
            bloom.write_to(&mut out)?;
            out.write_all(&records_end.to_be_bytes())?;
            out.write_all(&MAGIC.to_be_bytes())?;
            out.flush()?;
        }
        Self::open(path)
    }

    /// `None` - key not in this run.
    /// `Some(None)` - tombstoned in this run.
    /// `Some(Some(v))` - value present in this run.
    ///
    /// `check_bloom = false` skips the bloom probe and goes straight to the
    /// scan - used by [`BloomMode::Off`] to measure the optimisation's value.
    pub(crate) fn get(&self, key: &str, check_bloom: bool) -> Option<Option<Vec<u8>>> {
        if check_bloom && !self.bloom.might_contain(key) {
            return None;
        }
        let kb = key.as_bytes();
        // Records are sorted, so binary-search the offset index (O(log n)) rather
        // than scan from the file start. `binary_search_by` reads the key at each
        // probed offset - the index carries offsets only, not the keys.
        match self
            .offsets
            .binary_search_by(|&off| self.key_at(off).cmp(kb))
        {
            Ok(idx) => {
                let off = self.offsets[idx];
                let key_len =
                    u32::from_be_bytes(self.buf[off..off + 4].try_into().unwrap()) as usize;
                let mut p = off + 4 + key_len;
                let flag = self.buf[p];
                p += 1;
                let value_len = u32::from_be_bytes(self.buf[p..p + 4].try_into().unwrap()) as usize;
                p += 4;
                Some(if flag == FLAG_TOMBSTONE {
                    None
                } else {
                    Some(self.buf[p..p + value_len].to_vec())
                })
            }
            Err(_) => None,
        }
    }

    /// Records whose key is in `[lo, hi)` (either bound `None` = unbounded), in
    /// key order, as `(key, Option<value>)` - tombstones surface as `(key, None)`.
    /// Binary-searches the offset index to seek to `lo` (an `O(log n)` jump), then
    /// scans forward only across the window, breaking at `hi`.
    pub(crate) fn range(
        &self,
        lo: Option<&str>,
        hi: Option<&str>,
    ) -> Vec<(String, Option<Vec<u8>>)> {
        // Lower bound: first record whose key is >= lo. The offset index is in
        // key order, so `partition_point` finds it without touching earlier runs.
        let start = match lo {
            Some(l) => self
                .offsets
                .partition_point(|&off| self.key_at(off) < l.as_bytes()),
            None => 0,
        };
        let mut out = Vec::new();
        for &off in &self.offsets[start..] {
            let key_len = u32::from_be_bytes(self.buf[off..off + 4].try_into().unwrap()) as usize;
            let key_bytes = &self.buf[off + 4..off + 4 + key_len];
            if hi.map(|h| key_bytes >= h.as_bytes()).unwrap_or(false) {
                break;
            }
            let mut p = off + 4 + key_len;
            let flag = self.buf[p];
            p += 1;
            let value_len = u32::from_be_bytes(self.buf[p..p + 4].try_into().unwrap()) as usize;
            p += 4;
            let key = String::from_utf8_lossy(key_bytes).into_owned();
            let value = if flag == FLAG_TOMBSTONE {
                None
            } else {
                Some(self.buf[p..p + value_len].to_vec())
            };
            out.push((key, value));
        }
        out
    }

    #[allow(dead_code)]
    pub(crate) fn path(&self) -> &Path {
        &self.path
    }

    /// Every record in this run, in key order, as `(key, Option<value>)` -
    /// tombstones surface as `(key, None)`. Used by compaction to merge runs.
    pub(crate) fn entries(&self) -> Vec<(String, Option<Vec<u8>>)> {
        let mut out = Vec::new();
        let mut p = 0usize;
        while p < self.records_end {
            let key_len = u32::from_be_bytes(self.buf[p..p + 4].try_into().unwrap()) as usize;
            p += 4;
            let key = String::from_utf8_lossy(&self.buf[p..p + key_len]).into_owned();
            p += key_len;
            let flag = self.buf[p];
            p += 1;
            let value_len = u32::from_be_bytes(self.buf[p..p + 4].try_into().unwrap()) as usize;
            p += 4;
            let value = if flag == FLAG_TOMBSTONE {
                None
            } else {
                Some(self.buf[p..p + value_len].to_vec())
            };
            p += value_len;
            out.push((key, value));
        }
        out
    }
}
